kong/dao/migrations/helpers.lua

```diff
@@ -95,5 +99,205 @@ function _M.plugin_config_iterator(dao, plugin_name)
+        end
+      end
+
+      local question_marks = string.sub(string.rep("?, ", len), 1, -3)
+
+      local insert_cql = fmt("INSERT INTO %s (%s) VALUES (%s)",
+                             destination_table_def.name,
+                             table_concat(column_names, ", "),
+                             question_marks)
+      local _, err = coordinator:execute(insert_cql, values, nil, "write")
+      if err then
+        return nil, err
+      end
+    end
+  end
+end
+
+do
+
+  local function create_table(coordinator, table_def)
+
+    local partition_keys = table_def.partition_keys
+    local primary_key_cql = ""
+    if #partition_keys > 0 then
+      primary_key_cql = fmt(", PRIMARY KEY (%s)", table_concat(partition_keys, ", "))
+    end
+
+    local column_declarations = {}
+    local len = 0
+    for name, typ in pairs(table_def.columns) do
+      len = len + 1
+      column_declarations[len] = fmt("%s %s", name, typ)
+    end
+    local column_declarations_cql = table_concat(column_declarations, ", ")
+
+    local cql = fmt("CREATE TABLE %s(%s%s);",
+                    table_def.name,
+                    column_declarations_cql,
+                    primary_key_cql)
+    return coordinator:execute(cql, {}, nil, "write")
+  end
+
+
+  local function drop_table(coordinator, table_name)
+    local cql = fmt("DROP TABLE %s;", table_name)
+    return coordinator:execute(cql, {}, nil, "write")
+  end
+
+
+  local function get_columns_to_copy(table_structure)
+    local res = {}
+    for k, _ in pairs(table_structure.columns) do
+      res[k] = k
+    end
+    return res
+  end
+
+
+  local function create_aux_table_def(table_def)
+    local aux_table_def = utils.deep_copy(table_def)
+    aux_table_def.name = "copy_of_" .. table_def.name
+    aux_table_def.columns.partition = "text"
+    table.insert(aux_table_def.partition_keys, 1, "partition")
+    return aux_table_def
+  end
+
+  --[[
+    Add a new partition key called "partition" to the table specified by table_def.
+
+    table_def has the following structure:
+    { name    = "ssl_certificates",
+      columns = {
+        id         = "uuid",
+        cert       = "text",
+        key        = "text",
+        created_at = "timestamp",
+      },
+      partition_keys = { "id" },
+    }
+  --]]
+
+  function _M.cassandra.add_partition(dao, table_def)
+
+    local coordinator = dao.db:get_coordinator()
+
+    table_def = utils.deep_copy(table_def)
+
+    local aux_table_def = create_aux_table_def(table_def)
+    local columns_to_copy = get_columns_to_copy(table_def)
+    columns_to_copy.partition = function() return cassandra.text(table_def.name) end
+
+    local _, err = create_table(coordinator, aux_table_def)
+    if err then
+      return nil, err
+    end
+
+    local _, err = _M.copy_records(dao, table_def, aux_table_def, columns_to_copy)
+    if err then
+      return nil, err
+    end
+
+    local _, err = drop_table(coordinator, table_def.name)
```

Comment:
should we remove both tables at the last minute, to ensure none of them are lost in case one of the other operation fails?

Reply:
Also, we should probably make this `drop_table_if_exists`, to make this migration safe to re-run in case it previously failed halfway through. If it failed after a delete table but before finishing, it seems like it would never be able to fully complete (since all subsequent runs of the migration hit a roadblock trying to delete a non-existing table).